pkg/gencert/gencert.go

```diff
@@ -0,0 +1,179 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gencert
+
+import (
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	certificates "k8s.io/api/certificates/v1beta1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/cert"
+	"k8s.io/client-go/util/keyutil"
+	"time"
+)
+
+const (
+	// systemPrivilegedGroup is a superuser by default (i.e. bound to the cluster-admin ClusterRole).
+	systemPrivilegedGroup = "system:masters"
+	// certWaitInterval certificate request poll interval.
+	certWaitInterval = 1 * time.Second
+	// certWaitTimeout certificate request poll timeout.
+	certWaitTimeout = 20 * time.Second
+)
+
+// generateCSR generates a certificate signing request (CSR).
+func generateCSR() (*certificates.CertificateSigningRequest, []byte, error) {
+
+	// Generate a new private key.
+	pk, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to generate a new private key: %v", err)
```

Comment:
for the most part, I think `fmt.Errorf("generate key: %v", err)` is more concise and equally informative (applies everywhere).

(and if only one line in the function can return an error, then no need to add to it)